### PR TITLE
BUG: groupby.apply incorrectly dropping nan

### LINKED ISSUE #43205: Combination of groupby with dropna set to True and apply function not returning expected output
(Using pandas 1.3.1 and numpy 1.20.3)

This is a repetition of the SO question found [here](https://stackoverflow.com/questions/68918019/pandas-groupby-with-dropna-set-to-true/68918101?noredirect=1#comment121798992_68918101). (I did not have this issue previously when working with version 1.1.3).

In the following snippet:
```
import pandas as pd
import numpy as np

df = pd.DataFrame(
    {
        "a": [1, 2, 3, 4, 5, 6, 7, 8, 9], 
        "b": [1, np.nan, 1, np.nan, 2, 1, 2, np.nan, 1]
    }
)
df_again = df.groupby("b", dropna=False).apply(lambda x: x)
```
I was expecting df and df_again to be identical. But they are not:
```
df
   a    b
0  1  1.0
1  2  NaN
2  3  1.0
3  4  NaN
4  5  2.0
5  6  1.0
6  7  2.0
7  8  NaN
8  9  1.0

df_again
   a    b
0  1  1.0
2  3  1.0
4  5  2.0
5  6  1.0
6  7  2.0
8  9  1.0
```

Now, if I tweak slightly the lambda expression slightly to "see" what is going on by `df.groupby("b", dropna=False).apply(lambda x: print(x))` I can actually visualize that also the portion of the `df` where `b` was `NaN` was processed.

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #43205
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by debnathshoham** [comment]
since this is a regression, shouldn't this go to 1.3.3?

**Review by jreback — APPROVED** [review]
this looks fine
cc @rhshadrach

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1010,7 +1010,11 @@ def reset_identity(values):
 
         if not not_indexed_same:
             result = concat(values, axis=self.axis)
-            ax = self.filter(lambda x: True).axes[self.axis]
+            ax = (
+                self.filter(lambda x: True).axes[self.axis]
+                if self.dropna
+                else self._selected_obj._get_axis(self.axis)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:1016]: i dont have any reason to think it makes a difference, but big-picture ive been trying to figure out why we use _selected_obj in some places and _obj_with_exclusions in others so we can either a) document the difference and be consistent about it or b) get rid of one of them.

nothing to worry about for this PR, but if you're doing more work in the groupby code, id appreciate extra eyeballs trying to sort this out

[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1016]: The only difference I know is that much of the time, selected_obj includes the grouping column(s) whereas obj_with_exclusions does not. This is not true if a selection is performed after the groupby:

df = pd.DataFrame({'a': [1], 'b': [2], 'c': [3]})
print(df.groupby('a')._selected_obj)   # columns a, b, c
print(df.groupby('a')._obj_with_exclusions)  # columns b, c
print(df.groupby('a')['b']._selected_obj)  # Column b as a Series
print(df.groupby('a')['b']._obj_with_exclusions)  # Column b as a Series

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -1151,3 +1151,17 @@ def test_doctest_example2():
         {"B": [1.0, 0.0], "C": [2.0, 0.0]}, index=Index(["a", "b"], name="A")
     )
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("dropna", [True, False])
+def test_apply_dropna_with_indexed_same2(dropna):
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_apply.py:1157]: This test completely supersedes the other `test_apply_dropna_with_indexed` test; can you just apply the parametrization and the `expected` change to that test.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR - looks great.

File: pandas/tests/groupby/test_apply.py (hunk 2 above)
Comment (debnathshoham) [reply pandas/tests/groupby/test_apply.py]: Amended.
Let me know how this looks.

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by jreback** [comment]
thanks @debnathshoham

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Comment by jreback** [comment]
@debnathshoham if you can follow the instructions above and manually backport

**Pull request merged**